s3tests_boto3/functional/test_iam.py

```diff
@@ -0,0 +1,934 @@
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = client.delete_user_policy(PolicyName='AllowAccessPolicy2',
+                                         UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = client.get_user_policy(PolicyName='AllowAccessPolicy3',
+                                      UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    eq(response['PolicyName'], "AllowAccessPolicy3")
+    eq(response['PolicyDocument'], json.loads(policy_document_allow))
+
+    response = client.delete_user_policy(PolicyName='AllowAccessPolicy3',
+                                         UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
+
+@attr(resource='user-policy')
+@attr(method='s3 Actions')
+@attr(operation='Verify Allow Bucket Actions in user Policy')
+@attr(assertion='succeeds')
+@attr('user-policy')
+def test_allow_bucket_actions_in_user_policy():
+    client = get_iam_client()
+    s3_client = get_tenant_client()
+    bucket = get_new_bucket(client=s3_client)
+
+    bucket_listed = False
+
+    policy_document_allow = json.dumps(
+        {"Version": "2012-10-17",
+         "Statement": {
+             "Effect": "Allow",
+             "Action": ["s3:ListAllMyBuckets", "s3:DeleteBucket"],
+             "Resource": "arn:aws:s3:::*"}}
+    )
+    response = client.put_user_policy(PolicyDocument=policy_document_allow,
+                                      PolicyName='AllowAccessPolicy', UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
+    response = s3_client.list_buckets()
+    for index in range(len(response['Buckets'])):
+        if bucket == response['Buckets'][index]['Name']:
+            bucket_listed = True
+            break
+    if not bucket_listed:
+        raise AssertionError("bucket is not listed")
+
+    response = s3_client.delete_bucket(Bucket=bucket)
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 204)
+    response = s3_client.list_buckets()
+    bucket_listed = False
+    for index in range(len(response['Buckets'])):
+        if bucket == response['Buckets'][index]['Name']:
+            bucket_listed = True
+            break
+    if bucket_listed:
+        raise AssertionError("deleted bucket is getting listed")
+    response = client.delete_user_policy(PolicyName='AllowAccessPolicy',
+                                         UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
+
+@attr(resource='user-policy')
+@attr(method='s3 Actions')
+@attr(operation='Verify Deny Bucket Actions in user Policy')
+@attr(assertion='succeeds')
+@attr('user-policy')
+def test_deny_bucket_actions_in_user_policy():
+    client = get_iam_client()
+    s3_client = get_tenant_client()
+    bucket = get_new_bucket(client=s3_client)
```

Comment:
this test is ok, because the bucket being owned by tenant user, has all s3 actions allowed on it by default. The policy explicitly 'Denies' two actions and that is being tested here.